docs/new-spec.md

```diff
@@ -0,0 +1,347 @@
+    fn verify(_fingerprint: &Uint256, _signature: &Signature, _address: Address) -> bool;
+    fn web3<'ret, 'me: 'ret>(&'me self) -> RwLockReadGuardRef<'ret, Crypto, Web3Handle>;
+
+    // Async stuff
+    fn get_network_id(&self) -> impl Future<Item = u64, Error = Error>;
+    fn get_nonce(&self) -> impl Future<Item = Uint256, Error = Error>;
+    fn get_gas_price(&self) -> impl Future<Item = Uint256, Error = Error>;
+    /// Queries the network for current balance. This is different
+    /// from get_balance which keeps track of local balance to save
+    /// up on network calls.
+    ///
+    /// This function shouldn't be called every time. Ideally it should be
+    /// called once when initializing private key, or periodically to synchronise
+    /// local and network balance.
+    fn get_network_balance(&self) -> impl Future<Item = Uint256, Error = Error>;
+    /// Waits for an event on the network using the event name.
+    ///
+    /// * `event` - Event signature
+    /// * `topic` - First topic to filter out
+    fn wait_for_event(
+        &self,
+        event: &str,
+        topic: Option<[u8; 32]>,
+    ) -> impl Future<Item = Log, Error = Error>;
+    /// Broadcast a transaction on the network.
+    ///
+    /// * `action` - Defines a type of transaction
+    /// * `value` - How much wei to send
+    fn broadcast_transaction(
+        &self,
+        action: Action,
+        value: Uint256,
+    ) -> impl Future<Item = Uint256, Error = Error>;
+}
+
+## Storage
+
+This is a trait that implements a storage of Counterparties. It is left in tact for most of the part, only the interface is extracted as a trait.
+
+```rust
+// Open question: Naming of this trait?
+trait Storage {
+    pub fn get_all_counterparties(&self) -> impl Future<Item = Vec<Counterparty>, Error = Error>;
+    pub fn get_all_channel_managers_mut(
+        &self,
+    ) -> impl Future<Item = Vec<Guard<ChannelManager>>, Error = Error>;
+    pub fn get_channel(
+        &self,
+        k: Address,
+    ) -> impl Future<Item = Guard<ChannelManager>, Error = Error>;
+    fn init_data(
+        &self,
+        k: Counterparty,
+        v: ChannelManager,
+    ) -> impl Future<Item = (), Error = Error>;
+}
+```
+
+## TransportProtocol
+
+A trait that describes the node to node protocol described above in section [Transport layer](#transport-layer).
+
+_Open question: How do we name those traits and implementations? For now its example to illustrate the idea_
+
+```rust
+type ChannelId = [u8; 32];
+
+struct Channel {
+    channel_id: ChannelId,
+    address0: Address,
+    address1: Address,
+    balance0: Uint256,
+    balance1: Uint256,
+    // Rest of implementation details about the channel
+}
+
+trait TransportProtocol {
+    /// Proposes a channel and returns Signature after signing a fingerprint
+    fn propose(&mut self, channel: &Channel) -> impl Future<Item = Signature, Error = Error>;
+    /// Notifies about channel created
```

Comment:
note that this creates a future that will only return when the channel has been created. By the way doesn't this mean we need a create channel function to actually make channel created return ever? Because channel created is monitoring, not creating .

Reply:
ok not in scope of transport protocol because it's on chain.